LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/DeviceLoaderSynchronizer.cs

```diff
@@ -0,0 +1,330 @@
+
+namespace LoRaWan.NetworkServer
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using LoRaTools.Utils;
+    using Microsoft.Extensions.Logging;
+
+    /// <summary>
+    /// Represents a running task loading devices by devAddr
+    /// - Prevents querying the registry and loading twins multiple times
+    /// - Ensure that requests are queued by <see cref="LoRaDevice"/> in the order they come
+    /// </summary>
+    class DeviceLoaderSynchronizer : ILoRaDeviceRequestQueue
+    {
+        internal enum LoaderState
+        {
+            QueryingDevices,
+
+            CreatingDeviceInstances,
+
+            DispatchingQueuedItems,
+
+            Finished
+        }
+
+        private readonly LoRaDeviceAPIServiceBase loRaDeviceAPIService;
+        private readonly ILoRaDeviceFactory deviceFactory;
+        private readonly string devAddr;
+        private readonly DevEUIToLoRaDeviceDictionary destinationDictionary;
+        private readonly HashSet<ILoRaDeviceInitializer> initializers;
+        private readonly NetworkServerConfiguration configuration;
+        private readonly Task loading;
+        private volatile LoaderState state;
+        private volatile bool loadingDevicesFailed;
+        private object queueLock;
+        private volatile List<LoRaRequest> queuedRequests;
+
+        internal DeviceLoaderSynchronizer(
+            string devAddr,
+            LoRaDeviceAPIServiceBase loRaDeviceAPIService,
+            ILoRaDeviceFactory deviceFactory,
+            DevEUIToLoRaDeviceDictionary destinationDictionary,
+            HashSet<ILoRaDeviceInitializer> initializers,
+            NetworkServerConfiguration configuration,
+            Action<Task> continuationAction)
+        {
+            this.loRaDeviceAPIService = loRaDeviceAPIService;
+            this.deviceFactory = deviceFactory;
+            this.devAddr = devAddr;
+            this.destinationDictionary = destinationDictionary;
+            this.initializers = initializers;
+            this.configuration = configuration;
+            this.state = LoaderState.QueryingDevices;
+            this.loadingDevicesFailed = false;
+            this.queueLock = new object();
+            this.queuedRequests = new List<LoRaRequest>();
+            this.loading = this.Load().ContinueWith(continuationAction, TaskContinuationOptions.ExecuteSynchronously);
+        }
+
+        async Task Load()
+        {
+            try
+            {
+                // If device was not found, search in the device API, updating local cache
+                Logger.Log(this.devAddr, "querying the registry for device", LogLevel.Information);
+
+                SearchDevicesResult searchDeviceResult = null;
+                try
+                {
+                    searchDeviceResult = await this.loRaDeviceAPIService.SearchByDevAddrAsync(this.devAddr);
```

Comment:
Would be cool if you could explain to me, why a device can be found multiple times with this? I thought the devAddr is unique.

Reply:
Using ABP the user defines the DevAddr, thus there is no guarantee of uniqueness. With OTAA we generate the DevAddr, and there is a very very small chance of conflict.

To identify the correct device we need a combination of DevAddr + checking the payload  integrity (MIC) against the device network session key.